server/websession.py

```diff
@@ -0,0 +1,97 @@
+# code for this started by copying the bits of flask-session
+# https://github.com/mcrowson/flask-sessionstore-fork/blob/master/flask_sessionstore/sessions.py
+# that are relevant for database-backed storage, then simplifying for just our use case.
+
+import secrets
+from datetime import datetime, timezone
+
+from flask.sessions import SessionInterface, SessionMixin
+from werkzeug.datastructures import CallbackDict
+from .models import WebSession
+from .database import db_session
+from . import config
+
+
+class ArloSession(CallbackDict, SessionMixin):  # pylint: disable=too-many-ancestors
+    def __init__(self, sid=None, initial=None):
+        def on_update(self):
+            self.modified = True
+
+        CallbackDict.__init__(self, initial, on_update)
+        self.sid = sid
+        self.permanent = True
+        self.modified = False
+
+
+class ArloSessionInterface(SessionInterface):
+    def _generate_sid(self):
+        return secrets.token_urlsafe(50)
+
+    def open_session(self, app, request):
+        sid = request.cookies.get(app.session_cookie_name)
+        if not sid:
+            return ArloSession(sid=self._generate_sid())
+
+        saved_session = WebSession.query.filter_by(id=sid).first()
+
+        if saved_session:
+            return ArloSession(sid=sid, initial=saved_session.data)
+        else:
+            return ArloSession(sid=self._generate_sid())
+
+    def save_session(self, app, session, response):
+        if not session.modified:
+            return
+
+        # we're about to update the database, and we want to cancel any pending and uncommitted updates.
+        # because we use scoped_session, which maintains a thread-local sqlalchemy session,
+        # calling remove here tidies everything up, including potentially rolling back an uncommitted txn,
+        # and on next db_session call, we'll be using a new thread-local session.
+        db_session.remove()
+
+        saved_session = WebSession.query.filter_by(id=session.sid).first()
+
+        if saved_session:
+            saved_session.data = dict(session)
+        else:
+            new_session = WebSession(id=session.sid, data=dict(session))
+            db_session.add(new_session)
```

Comment:
Any worries about race conditions here? Do we need an atomic upsert? I haven't thought about it deeply but figured you might have

Reply:
great question! I did not think through this issue.

Thinking through it a bit, I don't think this is a problem. This only happens if a session has been deleted by the cleanup process (or if someone is messing with their cookie). If there's a race condition, it would be because two requests are happening at once... and whichever one completes second will overwrite the cookie from the first, and the first allocated session will be garbage collected a few minutes later.

So, good callout, but I think this is safe. Almost as important I don't know how we can do anything better, because we want to allocate a fresh session anyways.

Reply:
That reasoning makes sense to me!